### ISSUE TITLE
Question around set indexes

### ISSUE THREAD

**Issue by zombiemonkey** [yandex/ClickHouse#4716]
Hi there,

I'm a bit confused about set indexes and where they kick in. I've tried the examples in the tests folder and they work as expected however if I try to apply them in my case granules are never dropped. I've tried with larger values and more partitions (use toStartOfHour(ts) below) but set indexes never kick in. Am I mistaken in their use/purpose and have to wait for bloom indices or am I using them incorrectly?

Thanks!

```
SET allow_experimental_data_skipping_indices = 1;

DROP TABLE IF EXISTS test.idx_test;

CREATE TABLE test.idx_test (
                s_key UInt64,
                id UInt32,
                ts DateTime,
                value UInt64,
                INDEX s_key_idx (s_key) TYPE set(0) GRANULARITY 1000
) ENGINE = MergeTree
ORDER BY (id, ts)
SETTINGS index_granularity = 32;

INSERT INTO test.idx_test
SELECT
                cityHash64(id) AS s_key,
                number AS id,
                toDateTime(1551398400 + rand(1)%86400) AS ts,
                rand(2) AS value
FROM system.numbers LIMIT 100000;

SET send_logs_level = 'debug';

select * from test.idx_test where id = 3000 format PrettySpace;

[clickhouse-demo] 2019.03.17 22:01:10.564572 {44b0be5f-2a11-47f2-9ca5-abe862c3ab79} [ 55 ] <Debug> executeQuery: (from 127.0.0.1:34334) select * from test.idx_test where id = 3000 format PrettySpace;
[clickhouse-demo] 2019.03.17 22:01:10.566056 {44b0be5f-2a11-47f2-9ca5-abe862c3ab79} [ 55 ] <Debug> test.idx_test (SelectExecutor): Key condition: (column 0 in [3000, 3000])
[clickhouse-demo] 2019.03.17 22:01:10.566108 {44b0be5f-2a11-47f2-9ca5-abe862c3ab79} [ 55 ] <Debug> test.idx_test (SelectExecutor): Selected 1 parts by date, 1 parts by key, 1 marks to read from 1 ranges
[clickhouse-demo] 2019.03.17 22:01:10.566366 {44b0be5f-2a11-47f2-9ca5-abe862c3ab79} [ 55 ] <Debug> executeQuery: Query pipeline:

Expression
Expression
  Filter
   MergeTreeThread

               s_key     id                    ts        value

16286406272394286119   3000   2019-03-01 13:11:22   3080386888

[clickhouse-demo] 2019.03.17 22:01:10.568391 {44b0be5f-2a11-47f2-9ca5-abe862c3ab79} [ 55 ] <Information> executeQuery: Read 32 rows, 768.00 B in 0.004 sec., 8682 rows/sec., 203.51 KiB/sec.

[clickhouse-demo] 2019.03.17 22:01:10.568446 {44b0be5f-2a11-47f2-9ca5-abe862c3ab79} [ 55 ] <Debug> MemoryTracker: Peak memory usage (for query): 1.44 MiB.

1 rows in set. Elapsed: 0.004 sec.

select * from test.idx_test where s_key = 16286406272394286119 format PrettySpace;

[clickhouse-demo] 2019.03.17 22:01:35.857723 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> executeQuery: (from 127.0.0.1:34334) select * from test.idx_test where s_key = 16286406272394286119 format PrettySpace;
[clickhouse-demo] 2019.03.17 22:01:35.858430 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> InterpreterSelectQuery: MergeTreeWhereOptimizer: condition "s_key = 16286406272394286119" moved to PREWHERE
[clickhouse-demo] 2019.03.17 22:01:35.858879 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> test.idx_test (SelectExecutor): Key condition: unknown
[clickhouse-demo] 2019.03.17 22:01:35.859056 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> test.idx_test (SelectExecutor): Index `s_key_idx` has dropped 0 granules.
[clickhouse-demo] 2019.03.17 22:01:35.859096 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> test.idx_test (SelectExecutor): Selected 1 parts by date, 1 parts by key, 3125 marks to read from 1 ranges
[clickhouse-demo] 2019.03.17 22:01:35.859269 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> executeQuery: Query pipeline:
Expression
Expression
  MergeTreeThread

               s_key     id                    ts        value
16286406272394286119   3000   2019-03-01 13:11:22   3080386888

[clickhouse-demo] 2019.03.17 22:01:35.864492 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Information> executeQuery: Read 100000 rows, 781.27 KiB in 0.007 sec., 14970210 rows/sec., 114.22 MiB/sec.
[clickhouse-demo] 2019.03.17 22:01:35.864554 {2a2f2182-1abd-4037-834a-bb963b58afa6} [ 55 ] <Debug> MemoryTracker: Peak memory usage (for query): 3.99 MiB.

1 rows in set. Elapsed: 0.008 sec. Processed 100.00 thousand rows, 800.02 KB (12.81 million rows/s., 102.50 MB/s.)
```

**Comment by KochetovNicolai** [comment]
I think the reason is that `set(0)` is an empty set, so empty index was created. Try to use, for example, `set(100)`. 100 here is the max set size which is created per granule (so, if granule has more than 100 distinct values, set won't be created, and granule will never be dropped).

**Comment by zombiemonkey** [comment]
Ah ok - thanks Nicolai. Have it working now. The docs say that set(0) means no limit but that must be incorrect. Cheers!

**Comment by KochetovNicolai** [comment]
Well, it seems that set(0) should mean no limit, and it is fixed in #4640. Docs was correct.

**Comment by den-crane** [comment]
IMHO skip `INDEX s_key_idx (s_key) TYPE set(0) GRANULARITY 1000` just has no sense.
It makes a query slower and doubles the column stored data.


```
create table BX(
I Int64, 
S String,
INDEX Sx S TYPE set(0) GRANULARITY 1000
) Engine=MergeTree order by I;

insert into BX select number, toString(rand()) from numbers(10000000);
insert into BX values(45645645, '666');
select * from BX where S = '666'
┌────────I─┬─S───┐
│ 45645645 │ 666 │
└──────────┴─────┘
1 rows in set. Elapsed: 0.235 sec.
```

```
alter table BX drop index Sx;
select * from BX where S = '666'
┌────────I─┬─S───┐
│ 45645645 │ 666 │
└──────────┴─────┘
1 rows in set. Elapsed: 0.070 sec. Processed 10.00 million rows, 267.41 MB (143.10 million rows/s., 3.83 GB/s.)
```

**Comment by zombiemonkey** [comment]
@den-crane - I know :) I think you misunderstood that this was seeking clarity as to why set(0) was not working as documented. As KochetovNicolai pointed out - there was a bug fixed in #4640. 1000 was simply one of the values that was used for testing from a range of values and not the problem/issue WRT the behavior of set(0).
